prow/tide/tide.go

```diff
@@ -866,46 +865,105 @@ func (c *changedFilesAgent) prune() {
 	c.nextChangeCache = make(map[changeCacheKey][]string)
 }
 
-func (c *Controller) presubmitsByPull(sp *subpool) (map[int]sets.String, error) {
-	presubmits := make(map[int]sets.String, len(sp.prs))
-	record := func(num int, context string) {
-		if jobs, ok := presubmits[num]; ok {
-			jobs.Insert(context)
-		} else {
-			presubmits[num] = sets.NewString(context)
+// prJobRequirements describes prowjob requirements for a PR.
+type prJobRequirements struct {
+	// requiredPJContexts are the prowjob associated status contexts that Tide
+	// expects to be green before merge.
+	requiredPJContexts sets.String
+	// allowedPendingContexts are the contexts that are allowed to be pending without
+	// removing the PR from the pool. The PR is not mergeable until these turn green,
+	// but we don't want to thrash the pool when retesting.
+	// This set is a superset of requirePJContexts that includes the contexts of
+	// any parent jobs.
+	allowedPendingContexts sets.String
+	// toTrigger is the set of presubmit jobs to trigger for this PR if we decide
+	// to retest it. This is keyed by job name instead of context since the context
+	// may be unspecified if we are triggering a parent job that doesn't report.
+	toTrigger map[string]config.Presubmit
+}
+
+func (c *Controller) requirementsByPull(sp *subpool) (map[int]prJobRequirements, error) {
+	reqs := make(map[int]prJobRequirements, len(sp.prs))
+	for _, ps := range c.ca.Config().Presubmits[sp.org+"/"+sp.repo] {
+		// Does this job require a context or have kids that might need one?
+		if !ps.AlwaysRun && ps.RunIfChanged == "" {
+			continue // No
+		}
+		err := c.buildPRRequirements(reqs, sp.branch, sp.prs, sets.NewString(), &ps, []config.Presubmit{ps})
+		if err != nil {
+			return nil, err
 		}
 	}
+	return reqs, nil
+}
 
-	for _, ps := range c.ca.Config().Presubmits[sp.org+"/"+sp.repo] {
-		if !ps.ContextRequired() || !ps.RunsAgainstBranch(sp.branch) {
+func (c *Controller) buildPRRequirements(m map[int]prJobRequirements, branch string, prs []PullRequest, contextStack sets.String, rootJob *config.Presubmit, jobs []config.Presubmit) error {
```

Comment:
See my other responses. This doesn't check that the parent is required before processing children and it also doesn't distinguish the set of required contexts from the set of contexts that can be pending (required + parents of required).

I think if you add those 2 things and pass around a map instead of constructing and returning new slices every call you pretty much end up with what I have now?

Reply:
I'm suggesting the change because I have spent some number of minutes trying to figure out what this method is doing and still can't figure out if it is doing it correctly.

This method has 4 lists/maps/sets arguments as well as recursion and an inner function. Please break it apart into something comprehensible and testable.

Also I am suggesting a change to made the code easier to write and the config easier to read (marking the child job as required means the parent job will schedule). Currently I have to process the entire ancestry to determine if something will schedule. This is needlessly complicated.